File: java version/SudokuSolver/SudokuSolver.java
package SudokuSolver;

import java.util.Stack;

public class SudokuSolver
{
	public static abstract class BaseSolver
	{
		protected int[][] table = new int[9][9];
		
		protected boolean solved = false;
		protected boolean impossible = false;
		
		protected static class Position
		{
			int x = 0;
			int y = 0;
			boolean found = false;
			
			Position (int x, int y, boolean found)
			{
				this.x = x;
				this.y = y;
				this.found = found;
			}
		}
		
		protected boolean is_valid(int digit, int x, int y)
		{
			if (digit == 0)
				return true;
						
			for (int i = 0; i < 9; i++)
			{
				if (i != y && digit == this.table[i][x])
					return false;
			}
					
			for (int i = 0; i < 9; i++)
			{
				if (i != x && digit == this.table[y][i])
					return false;
			}
					
			int field_x = x / 3;
			int field_y = y / 3;
					
			for (int dy = 0; dy < 3; dy++)
			{
				for (int dx = 0; dx < 3; dx++)
				{
					int vx = field_x * 3 + dx;
					int vy = field_y * 3 + dy;
							
					if (vx == x && vy == y)
						continue;
								
					if (digit == this.table[vy][vx])
						return false;
				}
			}
			
			return true;
		}
		
		protected Position find_empty()
		{
			for (int y = 0; y < 9; y++)
			{
				for (int x = 0; x < 9; x++)
				{
					if (this.table[y][x] == 0)
					{
						return new Position(x, y, true);
					}
				}
			}
			
			return new Position(0, 0, false);
		}

		public void load_sudoku(String input)
		{
			for (int i = 0; i < 81; i++)
			{
				this.table[i / 9][i - (i / 9) * 9] = input.charAt(i) - '0';
			}
		}
		
		public void print_sudoku()
		{
			for (int i = 0; i < 9; i++)
			{
				if (i != 0 && i % 3 == 0)
				{
					for (int j = 0; j < 9; j++)
					{
						if (j != 0 && j % 3 == 0)
							System.out.print("+");
						System.out.print("-");
					}
					System.out.println();
				}
						
				for (int j = 0; j < 9; j++)
				{
					if (j != 0 && j % 3 == 0)
						System.out.print("|");
					System.out.print(this.table[i][j]);
				}
				System.out.println();;
			}
		}
		
		public void print_sudoku_str()
		{
			for (int i = 0; i < 9; i++)
			{
				for (int j = 0; j < 9; j++)
				{
					System.out.print(this.table[i][j]);
				}
			}
			System.out.println();
		}

		public abstract void solve();
	}
	
	public static class RecursiveSolver extends BaseSolver
	{
		public void solve()
		{
			Position free_pos = new Position(0, 0, false);
			
			free_pos = super.find_empty();
					
			if (!free_pos.found)
			{
				super.solved = true;
				System.out.println("Solved!");
				return;
			}
					
			for (int i = 1; i <= 9; i++)
			{
				if (super.is_valid(i, free_pos.x, free_pos.y))
				{
					super.table[free_pos.y][free_pos.x] = i;
					this.solve();
					if (!super.solved)
						super.table[free_pos.y][free_pos.x] = 0;
				}
			}
		}
	}
	
	public static class IterativeSolver extends BaseSolver
	{
		private static class Checkpoint {
			int digit;
			int x;
			int y;
			
			Checkpoint (int digit, int x, int y)
			{
				this.digit = digit;
				this.x = x;
				this.y = y;
			}
		}
		
		private Stack<Checkpoint> st = new Stack<>();

		public void solve()
		{
			int start_digit = 1;
			
			Position free_pos = new Position(0, 0, false);
					
			free_pos = super.find_empty();
					
			while (!super.solved && !super.impossible)
			{
				while (start_digit <= 9 && !super.is_valid(start_digit, free_pos.x, free_pos.y))
					start_digit++;
						
				if (start_digit <= 9)
				{
					super.table[free_pos.y][free_pos.x] = start_digit;
					this.st.push(new Checkpoint(start_digit, free_pos.x, free_pos.y));
					start_digit = 1;
					
					free_pos = super.find_empty();
							
					if (!free_pos.found)
					{
						super.solved = true;
						Checkpoint rem;
						while (!this.st.empty())
							rem = this.st.pop();
						System.out.println("Solved!");
					}
				}
				else
				{
					if (this.st.empty())
					{
						super.impossible = true;
						System.out.println("Impossible!");
					}
					else
					{
						Checkpoint rev = this.st.pop();
						free_pos.x = rev.x;
						free_pos.y = rev.y;
						start_digit = rev.digit + 1;
						super.table[free_pos.y][free_pos.x] = 0;
					}
				}
			}
		}
	}
}

File: kotlin-version/src/SudokuSolver/SudokuSolver.kt
package SudokuSolver

import java.util.Stack

class SudokuSolver
{
    public abstract class BaseSolver
    {
        protected var table: Array<Array<Int>> = Array(9){Array(9){0}}

        protected var solved: Boolean = false
        protected var impossible: Boolean = false

        protected class Position(
            var x: Int = 0,
            var y: Int = 0,
            var found: Boolean = false
        )

        protected fun is_valid(digit: Int, x: Int, y: Int): Boolean
        {
            if (digit == 0)
                return true

            for (i in 0..8)
                if (i != y && digit == this.table[i][x])
                    return false

            for (i in 0..8)
                if (i != x && digit == this.table[y][i])
                    return false

            var field_x: Int = x / 3
            var field_y: Int = y / 3

            for (dy in 0..2)
            {
                for (dx in 0..2)
                {
                    var vx: Int = field_x * 3 + dx
                    var vy: Int = field_y * 3 + dy

                    if (vx == x && vy == y)
                        continue

                    if (digit == this.table[vy][vx])
                        return false
                }
            }

            return true
        }

        protected fun find_empty(): Position
        {
            for (y in 0..8)
                for (x in 0..8)
                    if (this.table[y][x] == 0)
                        return Position(x, y, true)

            return Position(0, 0, false)
        }

        public fun load_sudoku(input: String)
        {
            for (i in 0..80)
                this.table[i / 9][i - (i / 9) * 9] = input[i] - '0'
        }

        public fun print_sudoku()
        {
            for (i in 0..8)
            {
                if (i != 0 && i % 3 == 0)
                {
                    for (j in 0..8)
                    {
                        if (j != 0 && j % 3 == 0)
                            print("+")
                        print("-")
                    }
                    println()
                }

                for (j in 0..8)
                {
                    if (j != 0 && j % 3 == 0)
                        print("|")
                    print(this.table[i][j])
                }
                println()
            }
        }

        public fun print_sudoku_str()
        {
            for (i in 0..8)
                for (j in 0..8)
                    print(this.table[i][j])
            println()
        }

        public abstract fun solve()
    }

    public class RecursiveSolver : BaseSolver()
    {
        public override fun solve()
        {
            var free_pos: Position = super.find_empty()

            if (!free_pos.found)
            {
                super.solved = true
                println("Solved!")
                return
            }

            for (i in 1..9)
            {
                if (super.is_valid(i, free_pos.x, free_pos.y))
                {
                    super.table[free_pos.y][free_pos.x] = i
                    this.solve()
                    if (!super.solved)
                        super.table[free_pos.y][free_pos.x] = 0
                }
            }
        }
    }

    public class IterativeSolver : BaseSolver()
    {
        private class Checkpoint (
            var digit: Int,
            var x: Int,
            var y: Int
        )

        private var st: Stack<Checkpoint> = Stack<Checkpoint>()

        public override fun solve()
        {
            var start_digit: Int = 1
            var free_pos: Position = Position(0, 0, false)
            free_pos = super.find_empty()

            while (!super.solved && !super.impossible)
            {
                while (start_digit <= 9 && !super.is_valid(start_digit, free_pos.x, free_pos.y))
                    start_digit++

                if (start_digit <= 9)
                {
                    super.table[free_pos.y][free_pos.x] = start_digit
                    this.st.push(Checkpoint(start_digit, free_pos.x, free_pos.y))
                    start_digit = 1

                    free_pos = super.find_empty()

                    if (!free_pos.found)
                    {
                        super.solved = true
                        this.st.clear()
                        println("Solved!")
                    }
                }
                else
                {
                    if (this.st.empty())
                    {
                        super.impossible = true
                        println("Impossible!")
                    }
                    else
                    {
                        var rev: Checkpoint = this.st.pop()
                        free_pos.x = rev.x
                        free_pos.y = rev.y
                        start_digit = rev.digit + 1
                        super.table[free_pos.y][free_pos.x] = 0
                    }
                }
            }
        }
    }
}